### PR TITLE
orchagent: Refacotring

### PR DESCRIPTION
**Pull request by stcheng**
- RouteOrch depends on NeighOrch
- Move m_syncdNextHops to NeighOrch class

### CODE REVIEW

File: orchagent/neighorch.h [hunk 1]
```diff
@@ -16,23 +20,48 @@ struct NeighborEntry
...
+    NeighOrch(DBConnector *db, string tableName, PortsOrch *portsOrch) :
+        Orch(db, tableName),
+        m_portsOrch(portsOrch),
+        m_nextHopGroupCount(0) {};
+
+    bool contains(IpAddress);
+    bool contains(IpAddresses);
+
+    bool addNextHopGroup(IpAddresses);
+    bool removeNextHopGroup(IpAddresses);
```
Comment (lguohan) [inline orchagent/neighorch.h]: I can see why nexthop management is moved to neigh orch, but do not fully understand why the nhg logic also moves to here? nhg is dynamically created/removed based on the route input, so probably still better be in route orch. is there any particular reason for this move?

File: orchagent/neighorch.h [hunk 2]
```diff
@@ -16,23 +20,48 @@ struct NeighborEntry
...
 {
 public:
-    NeighOrch(DBConnector *db, string tableName, PortsOrch *portsOrch, RouteOrch *routeOrch) :
-        Orch(db, tableName), m_portsOrch(portsOrch), m_routeOrch(routeOrch) {};
+    NeighOrch(DBConnector *db, string tableName, PortsOrch *portsOrch) :
+        Orch(db, tableName),
+        m_portsOrch(portsOrch),
+        m_nextHopGroupCount(0) {};
+
+    bool contains(IpAddress);
```
Comment (lguohan) [inline orchagent/neighorch.h]: -> has?

File: orchagent/routeorch.cpp [hunk 3]
```diff
@@ -297,65 +182,20 @@ bool RouteOrch::addRoute(IpPrefix ipPrefix, IpAddresses nextHops)
...
             {
                 /* Randomly pick an address from the address set. */
+                set<IpAddress> next_hop_set = nextHops.getIpAddresses();
                 auto it = next_hop_set.begin();
                 advance(it, rand() % next_hop_set.size());
 
-                /*
-                 * Set the route's temporary next hop to be the randomly picked one.
-                 */
+                /* Set the route's temporary next hop to be the randomly picked one. */
```
Comment (lguohan) [inline orchagent/routeorch.cpp]: it is possible that the choosed next hop is not in the neighbor list yet. then you are going to fail to add this tmp route. can we choose one that is already in the neighbor list?

File: orchagent/neighorch.h (hunk 1 above)
Comment (stcheng) [inline orchagent/neighorch.h]: The next hop group goes with next hop as it is just a bunch of neighbors forming a next hop group. As and the m_syncdNextHops data structure is in this class, it makes sense to move addNextHopGroup/removeNextHopGroup method to this class. So all the next hop (group) concepts are here.

**Comment by stcheng** [comment]
Updated.

**Comment by stcheng** [comment]
Will update the pull request with separating next hop with next hop group.

**Comment by lguohan** [comment]
what updated?

File: orchagent/neighorch.cpp [hunk 4]
```diff
@@ -226,7 +291,7 @@ bool NeighOrch::removeNeighbor(NeighborEntry neighborEntry)
     }
 
     m_syncdNeighbors.erase(neighborEntry);
-    m_routeOrch->removeNextHopEntry(ip_address);
+    removeNextHop(ip_address);
```
Comment (lguohan) [inline orchagent/neighorch.cpp]: line 279-290, why do we need to create the neighbor entry if remove is not successful?

Comment (stcheng) [inline orchagent/neighorch.cpp]: updated.

**Comment by lguohan** [comment]
ack

**Pull request merged**